base/compiler/ssair/inlining.jl

```diff
@@ -951,15 +955,10 @@ function analyze_method!(match::MethodMatch, argtypes::Vector{Any},
         (allow_typevars && !may_have_fcalls(match.method)) || return nothing
     end
 
-    # See if there exists a specialization for this method signature
-    mi = specialize_method(match; preexisting=true) # Union{Nothing, MethodInstance}
-    if mi === nothing
-        et = InliningEdgeTracker(state.et, invokesig)
-        effects = info_effects(nothing, match, state)
-        return compileable_specialization(match, effects, et, info;
-            compilesig_invokes=OptimizationParams(state.interp).compilesig_invokes)
-    end
-
+    # Get the specialization for this method signature
+    # (later we will decide what to do with it)
+    mi = specialize_method(match) # Union{Nothing, MethodInstance}
```

Comment:
Any reason to drop the `preexisting` option here?
```suggestion
    mi = specialize_method(match; preexisting=true) # Union{Nothing, MethodInstance}
```

Reply:
yeah, we require it to make the edge correct now

Reply:
Shouldn't abstractinterpret have inferred the exactly same edge already, right? We should be able to look up it at this point.

Reply:
If we force fresh specialization here, `mi` never be `nothing` so we can drop the check at the next line.

Reply:
Not precisely, this is right where we know inference might have given up on this edge and decided to ignore it, but optimization still wants to generate an `:invoke` so that it can resume here later at runtime it required.

I guess that makes sense that specialize_method shouldn't fail here. So would you recommend a type-assert instead of the return branch?

Reply:
Okay.

> So would you recommend a type-assert instead of the return branch?

I think the call of `specialize_method(::MethodMatch)` is inferred to return `MethodInstance` precisely on your branch.
So the return branch is dead already. We'd probably just want to delete it.